src/t8_data/t8_element_array_iterator.hxx

```diff
@@ -0,0 +1,191 @@
+  t8_element_array_iterator&
+  operator= (t8_element_array_iterator&& other)
+    = default;
+
+  /* Destructor */
+  ~t8_element_array_iterator () = default;
+
+  /* Dereferencing operator of the iterator wrapper, returning a reference to const of the t8_element_t-pointer to the serialized bytes representing the actual serialized t8_element_t pointer. */
+  const reference
+  operator* ()
+  {
+    T8_ASSERT (index_ >= 0 && static_cast<size_t> (index_) < array_->elem_count);
+    dref_element_ = static_cast<t8_element_t*> (t8_sc_array_index_locidx (array_, index_));
+    return dref_element_;
+  };
+
+  /* Pre- and Postfix increment */
+  t8_element_array_iterator&
+  operator++ ()
+  {
+    ++index_;
+    return *this;
+  };
+  t8_element_array_iterator
+  operator++ (int)
+  {
+    t8_element_array_iterator tmp_iterator (*this);
+    ++(*this);
+    return tmp_iterator;
+  };
+
+  /* Pre- and Postfix decrement */
+  t8_element_array_iterator&
+  operator-- ()
+  {
+    --index_;
+    return *this;
+  };
+  t8_element_array_iterator
+  operator-- (int)
+  {
+    t8_element_array_iterator tmp_iterator (*this);
+    --(*this);
+    return tmp_iterator;
+  };
+
+  /* Return the index within the array the iterator currently points to [0,...,size]. */
+  t8_locidx_t
+  GetArrayIndex () const
+  {
+    return index_;
+  };
+
+  /* Compute the linear id at a given level for the element the iterator points to. */
+  t8_linearidx_t
+  GetLinearIDAtLevel (const int level)
+  {
+    T8_ASSERT (index_ >= 0 && static_cast<size_t> (index_) < array_->elem_count);
+    dref_element_ = static_cast<t8_element_t*> (t8_sc_array_index_locidx (array_, index_));
+    return scheme_->t8_element_get_linear_id (dref_element_, level);
+  };
+
+  /* Comparison operators */
+  friend bool
+  operator== (const t8_element_array_iterator& iter1, const t8_element_array_iterator& iter2)
+  {
+    return (iter1.array_->array == iter2.array_->array && iter1.index_ == iter2.index_);
+  };
+  friend bool
+  operator!= (const t8_element_array_iterator& iter1, const t8_element_array_iterator& iter2)
+  {
+    return (iter1.array_->array != iter2.array_->array || iter1.index_ != iter2.index_);
+  };
+
+  /* Arithmetic assignment operators */
+  t8_element_array_iterator&
+  operator+= (const difference_type n)
+  {
+    index_ += n;
+    return *this;
+  }
+  t8_element_array_iterator&
+  operator-= (const difference_type n)
+  {
+    index_ -= n;
+    T8_ASSERT (index_ < 0);
+    return *this;
+  }
+
+ private:
+  const t8_eclass_scheme_c* scheme_; /*!< The scheme of the elements residing within the array. */
+  const sc_array_t* array_;          /*!< A pointer to the actual serialized array of element pointers. */
+  t8_locidx_t index_ { 0 };          /*!< The index the iterator currently points to. */
+  t8_element_t* dref_element_; /*!< A helper variable for de-serializing the actual element pointers in the array. */
```

Comment:
Yes, this was a point I did not like in the implementation but somehow I wanted to return a (const) reference from the dereferencing operator and this was not possible, because due to casting of the underlying bytes we created a temporary object to which we cannot return a reference, therefore, there was the need to store  a t8_element_t pointer as a member of the class. However, now the return type of the dereferencing operator changed to a value_type and therefore, we could get rid of the intermediate process and the class member holding the t8_element_t pointer. I think now it is a nice solution.